File: lib/src/main/java/wang/mycroft/lib/base/LoadingDialogHelper.java
package wang.mycroft.lib.base;

import android.app.Dialog;
import android.os.Looper;

import androidx.annotation.NonNull;
import androidx.lifecycle.DefaultLifecycleObserver;
import androidx.lifecycle.LifecycleOwner;

import static com.blankj.utilcode.util.Utils.runOnUiThread;

/**
 * 加载dialog的帮助类
 *
 * @author mycroft
 */
public final class LoadingDialogHelper implements DefaultLifecycleObserver {

    private DialogCreator dialogCreator;

    public LoadingDialogHelper(@NonNull LifecycleOwner lifecycleOwner, @NonNull DialogCreator dialogCreator) {
        lifecycleOwner.getLifecycle().addObserver(this);
        this.dialogCreator = dialogCreator;
    }

    private Dialog loadingDialog;

    /**
     * 显示通用的加载{@link Dialog}, 默认cancelable=false
     */
    public final void showLoadingDialog() {
        showLoadingDialog(false);
    }

    /**
     * 显示通用的加载{@link Dialog}
     *
     * @param cancelable 是否允许点击空白处取消
     */
    public final void showLoadingDialog(boolean cancelable) {
        if (loadingDialog == null) {
            loadingDialog = dialogCreator.createLoadingDialog();
        }
        loadingDialog.setCancelable(cancelable);

        if (Looper.myLooper() == Looper.getMainLooper()) {
            loadingDialog.show();
        } else {
            runOnUiThread(loadingDialog::show);
        }
    }

    /**
     * 隐藏加载Dialog
     */
    public final void hideLoadingDialog() {
        if (loadingDialog != null && loadingDialog.isShowing()) {
            if (Looper.myLooper() == Looper.getMainLooper()) {
                loadingDialog.cancel();
            } else {
                runOnUiThread(loadingDialog::cancel);
            }
        }
    }

    @Override
    public void onDestroy(@NonNull LifecycleOwner owner) {
        if (loadingDialog != null) {
            if (loadingDialog.isShowing()) {
                loadingDialog.cancel();
            }
            loadingDialog = null;
        }
    }

    public interface DialogCreator {
        /**
         * create loading dialog
         *
         * @return dialog
         */
        Dialog createLoadingDialog();
    }
}

File: lib/src/main/java/wang/mycroft/lib/base/LoadingDialogHelper.kt
package wang.mycroft.lib.base

import android.app.Dialog
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.LifecycleOwner
import com.blankj.utilcode.util.Utils

/**
 * 加载dialog的帮助类
 *
 * @author Mycroft Wong
 * @date 2019年11月15日
 */
class LoadingDialogHelper(
    lifecycleOwner: LifecycleOwner, private val dialogCreator: DialogCreator
) : DefaultLifecycleObserver {

    init {
        lifecycleOwner.lifecycle.addObserver(this)
    }

    private var loadingDialog: Dialog? = null

    /**
     * 显示通用的加载[Dialog], 默认cancelable=false
     */
    fun showLoadingDialog() {
        showLoadingDialog(false)
    }

    /**
     * 显示通用的加载[Dialog]
     *
     * @param cancelable 是否允许点击空白处取消
     */
    fun showLoadingDialog(cancelable: Boolean) {
        if (loadingDialog == null) {
            loadingDialog = dialogCreator.createLoadingDialog()
        }
        loadingDialog!!.setCancelable(cancelable)
        Utils.runOnUiThread { loadingDialog!!.show() }
    }

    /**
     * 隐藏加载Dialog
     */
    fun hideLoadingDialog() {
        if (loadingDialog?.isShowing == true) {
            Utils.runOnUiThread { loadingDialog!!.cancel() }
        }
    }

    override fun onDestroy(owner: LifecycleOwner) {
        if (loadingDialog != null) {
            if (loadingDialog!!.isShowing) {
                loadingDialog!!.cancel()
            }
            loadingDialog = null
        }
    }
}